File: impl/python/src/img_drv/nix/emit.py
"""Print an AST as valid Nix SOURCE. This is the transpiler.

The arrow ``EXPR -> .nix`` of ``docs/architecture.md``.

Fully parenthesised, on purpose. Extra parentheses cannot change a parse, so
the emitted file is correct BY CONSTRUCTION rather than correct if the
precedence table was transcribed properly. ``docs/nix-internals.md`` records two
levels that surprise everyone (``!`` binds looser than ``+``, ``//`` tighter
than the comparisons); a minimal-parenthesis printer has to get both right and
this one cannot get them wrong.

Making the output pretty is a separate, later job with its own test: emit,
re-parse, compare the ASTs.
"""

from __future__ import annotations

from . import ast

__all__ = ["to_nix"]

_ESCAPES = {
    '"': '\\"',
    "\\": "\\\\",
    "\n": "\\n",
    "\r": "\\r",
    "\t": "\\t",
    # `${` in a literal must be escaped or Nix reads an interpolation.
    "$": "\\$",
}


def _path_literal(p: str) -> str:
    """Write a path so it lexes back as a PATH.

    The ROOT path cannot be written as a bare separator: that lexes as the
    DIVISION OPERATOR, so the emitted source does not parse at all. Nix source
    spells it ``/.``, which is what nixpkgs itself writes.

    The differential printer prints it as ``/``, because Nix does. The two
    printers disagree on purpose, and only the round-trip law could see it.
    """
    return "/." if p == "/" else p


def _float_literal(f: float) -> str:
    """Write a float so it lexes back as a FLOAT.

    ``%g`` prints ``1.0`` as ``1``, which re-parses as an INTEGER, and Nix
    distinguishes the two: ``1 / 2`` is 0 for integers and 0.5 for floats. So
    the transpiler was silently able to change arithmetic. Found by the
    round-trip law, not by the eleven conformance intents, none of which
    contains a float.

    The differential printer must NOT do this either: ``nix-instantiate
    --parse`` really does print ``1.0`` as ``1``, and that is pinned by a
    vector.
    """
    text = f"{f:g}"
    if any(c in text for c in ".eEn"):
        return text
    return text + ".0"


def _escape(text: str) -> str:
    return "".join(_ESCAPES.get(c, c) for c in text)


def _parts(parts: tuple[ast.Part, ...]) -> str:
    out = []
    for p in parts:
        if isinstance(p, ast.Lit):
            out.append(_escape(p.text))
        else:
            out.append("${" + _expr(p.expr) + "}")
    return '"' + "".join(out) + '"'


def _attr(a: ast.Attr) -> str:
    """Render one attribute-path component.

    A string-named attribute is emitted as a STRING, not wrapped in an
    interpolation. Writing ``${"a${x}b"}`` is valid Nix denoting the same
    attribute, but it re-parses as a DYNAMIC name rather than a string one, and
    Nix keeps those apart. The differential printer collapses them, so nothing
    but the round-trip law could see it.
    """
    if isinstance(a, ast.Id):
        return a.name
    if isinstance(a, ast.DynAttr):
        return "${" + _expr(a.expr) + "}"
    if len(a.parts) == 1 and isinstance(a.parts[0], ast.Lit):
        return '"' + _escape(a.parts[0].text) + '"'
    return _parts(a.parts)


def _attrpath(path: ast.AttrPath) -> str:
    return ".".join(_attr(a) for a in path)


def _bind(b: ast.Binding) -> str:
    if isinstance(b, ast.Bind):
        return f"{_attrpath(b.path)} = {_expr(b.value)}; "
    source = f" ({_expr(b.source)})" if b.source is not None else ""
    names = "".join(" " + _attr(a) for a in b.attrs)
    return f"inherit{source}{names}; "


def _pattern(p: ast.Pattern) -> str:
    if isinstance(p, ast.Pvar):
        return p.name
    formals = ", ".join(
        name if default is None else f"{name} ? {_expr(default)}"
        for name, default in p.formals
    )
    if p.ellipsis:
        formals = "..." if not p.formals else formals + ", ..."
    alias = f" @ {p.alias}" if p.alias is not None else ""
    return "{ " + formals + " }" + alias


def _expr(e: ast.Expr) -> str:
    match e:
        case ast.Int():
            return str(e.value)
        case ast.Float():
            return _float_literal(e.value)
        case ast.Var():
            return e.name
        case ast.Path():
            return _path_literal(e.text)
        case ast.SearchPath():
            return f"<{e.text}>"
        case ast.Uri():
            return '"' + _escape(e.text) + '"'
        case ast.Str() | ast.IndStr():
            return _parts(e.parts)
        case ast.PathInterp():
            # The leading path is absolute after parsing, so this emits an
            # absolute interpolated path: valid Nix, and the same file.
            out: list[str] = []
            for part in e.parts:
                if isinstance(part, ast.Lit):
                    out.append(part.text)
                elif isinstance(part.expr, ast.Path):
                    out.append(part.expr.text)
                else:
                    out.append("${" + _expr(part.expr) + "}")
            return "".join(out)
        case ast.Not():
            return f"(!{_expr(e.expr)})"
        case ast.Neg():
            return f"(-{_expr(e.expr)})"
        case ast.Op():
            return f"({_expr(e.left)} {e.op.value} {_expr(e.right)})"
        case ast.HasAttr():
            return f"({_expr(e.expr)} ? {_attrpath(e.path)})"
        case ast.Apply():
            return f"({_expr(e.func)} {_expr(e.arg)})"
        case ast.Select():
            default = "" if e.default is None else f" or {_expr(e.default)}"
            return f"({_expr(e.expr)}.{_attrpath(e.path)}{default})"
        case ast.Lambda():
            return f"({_pattern(e.pattern)}: {_expr(e.body)})"
        case ast.List():
            return "[ " + "".join(_expr(i) + " " for i in e.items) + "]"
        case ast.AttrSet():
            rec = "rec " if e.recursive else ""
            return rec + "{ " + "".join(_bind(b) for b in e.binds) + "}"
        case ast.Let():
            return (
                "(let "
                + "".join(_bind(b) for b in e.binds)
                + f"in {_expr(e.body)})"
            )
        case ast.With():
            return f"(with {_expr(e.scope)}; {_expr(e.body)})"
        case ast.Assert():
            return f"(assert {_expr(e.condition)}; {_expr(e.body)})"
        case ast.If():
            return (
                f"(if {_expr(e.condition)} then {_expr(e.then)} "
                f"else {_expr(e.otherwise)})"
            )


def to_nix(e: ast.Expr) -> str:
    """The AST as Nix source."""
    return _expr(e)

File: impl/go/nix/emit.go
package nix

import (
	"fmt"
	"strconv"
	"strings"
)

// Print an AST as valid Nix SOURCE. This is the transpiler.
//
// The arrow EXPR -> .nix of docs/architecture.md.
//
// Fully parenthesised, on purpose. Extra parentheses cannot change a parse, so
// the emitted file is correct BY CONSTRUCTION rather than correct if the
// precedence table was transcribed properly. docs/nix-internals.md records two
// levels that surprise everyone (! binds looser than +, // tighter than the
// comparisons); a minimal-parenthesis printer has to get both right and this
// one cannot get them wrong.
//
// Making the output pretty is a separate, later job with its own test: emit,
// re-parse, compare the ASTs.

// escaper rewrites the characters that cannot appear raw in a Nix string.
//
// The $ is the one that matters and the one that is easy to miss: Nix reads
// ${ as the start of an antiquotation, so a literal dollar must be escaped or
// the file either fails to parse or, worse, parses with a different meaning.
var escaper = strings.NewReplacer(
	`"`, `\"`,
	`\`, `\\`,
	"\n", `\n`,
	"\r", `\r`,
	"\t", `\t`,
	`$`, `\$`,
)

func escape(s string) string { return escaper.Replace(s) }

func writeParts(b *strings.Builder, parts []Part) {
	b.WriteString(`"`)
	for _, p := range parts {
		switch p := p.(type) {
		case Lit:
			b.WriteString(escape(p.Text))
		case Anti:
			b.WriteString("${")
			writeExpr(b, p.Expr)
			b.WriteString("}")
		default:
			panic(fmt.Sprintf("unknown string part %T", p))
		}
	}
	b.WriteString(`"`)
}

func writeAttr(b *strings.Builder, a Attr) {
	switch a := a.(type) {
	case ID:
		b.WriteString(a.Name)
	case StrAttr:
		if len(a.Parts) == 1 {
			if lit, ok := a.Parts[0].(Lit); ok {
				b.WriteString(`"` + escape(lit.Text) + `"`)
				return
			}
		}
		b.WriteString("${")
		writeParts(b, a.Parts)
		b.WriteString("}")
	default:
		panic(fmt.Sprintf("unknown attribute %T", a))
	}
}

func writeAttrPath(b *strings.Builder, path AttrPath) {
	for i, a := range path {
		if i > 0 {
			b.WriteString(".")
		}
		writeAttr(b, a)
	}
}

func writeBinding(b *strings.Builder, bind Binding) {
	switch bind := bind.(type) {
	case Bind:
		writeAttrPath(b, bind.Path)
		b.WriteString(" = ")
		writeExpr(b, bind.Value)
		b.WriteString("; ")
	case Inherit:
		b.WriteString("inherit")
		if bind.Source != nil {
			b.WriteString(" (")
			writeExpr(b, bind.Source)
			b.WriteString(")")
		}
		for _, a := range bind.Attrs {
			b.WriteString(" ")
			writeAttr(b, a)
		}
		b.WriteString("; ")
	default:
		panic(fmt.Sprintf("unknown binding %T", bind))
	}
}

func writePattern(b *strings.Builder, p Pattern) {
	switch p := p.(type) {
	case PVar:
		b.WriteString(p.Name)
	case PSet:
		b.WriteString("{ ")
		for i, f := range p.Formals {
			if i > 0 {
				b.WriteString(", ")
			}
			b.WriteString(f.Name)
			if f.Default != nil {
				b.WriteString(" ? ")
				writeExpr(b, f.Default)
			}
		}
		if p.Ellipsis {
			if len(p.Formals) == 0 {
				b.WriteString("...")
			} else {
				b.WriteString(", ...")
			}
		}
		b.WriteString(" }")
		if p.Alias != "" {
			b.WriteString(" @ " + p.Alias)
		}
	default:
		panic(fmt.Sprintf("unknown pattern %T", p))
	}
}

// writeExpr is the type switch a variant would make exhaustive.
//
// Go does not check that every case is present, so the default panics rather
// than silently emitting nothing. A missing case is then a loud failure in the
// first test that reaches it, which is the best this encoding allows.
func writeExpr(b *strings.Builder, e Expr) {
	switch e := e.(type) {
	case Int:
		b.WriteString(strconv.FormatInt(e.Value, 10))
	case Float:
		b.WriteString(strconv.FormatFloat(e.Value, 'g', -1, 64))
	case Var:
		b.WriteString(e.Name)
	case PathLit:
		b.WriteString(e.Text)
	case SearchPath:
		b.WriteString("<" + e.Text + ">")
	case URI:
		b.WriteString(`"` + escape(e.Text) + `"`)
	case Str:
		writeParts(b, e.Parts)
	case IndStr:
		writeParts(b, e.Parts)
	case Not:
		b.WriteString("(!")
		writeExpr(b, e.Expr)
		b.WriteString(")")
	case Neg:
		b.WriteString("(-")
		writeExpr(b, e.Expr)
		b.WriteString(")")
	case BinOp:
		b.WriteString("(")
		writeExpr(b, e.Left)
		b.WriteString(" " + e.Op.Text() + " ")
		writeExpr(b, e.Right)
		b.WriteString(")")
	case HasAttr:
		b.WriteString("(")
		writeExpr(b, e.Expr)
		b.WriteString(" ? ")
		writeAttrPath(b, e.Path)
		b.WriteString(")")
	case Apply:
		b.WriteString("(")
		writeExpr(b, e.Func)
		b.WriteString(" ")
		writeExpr(b, e.Arg)
		b.WriteString(")")
	case Select:
		b.WriteString("(")
		writeExpr(b, e.Expr)
		b.WriteString(".")
		writeAttrPath(b, e.Path)
		if e.Default != nil {
			b.WriteString(" or ")
			writeExpr(b, e.Default)
		}
		b.WriteString(")")
	case Lambda:
		b.WriteString("(")
		writePattern(b, e.Pattern)
		b.WriteString(": ")
		writeExpr(b, e.Body)
		b.WriteString(")")
	case List:
		b.WriteString("[ ")
		for _, i := range e.Items {
			writeExpr(b, i)
			b.WriteString(" ")
		}
		b.WriteString("]")
	case AttrSet:
		if e.Recursive {
			b.WriteString("rec ")
		}
		b.WriteString("{ ")
		for _, bind := range e.Binds {
			writeBinding(b, bind)
		}
		b.WriteString("}")
	case Let:
		b.WriteString("(let ")
		for _, bind := range e.Binds {
			writeBinding(b, bind)
		}
		b.WriteString("in ")
		writeExpr(b, e.Body)
		b.WriteString(")")
	case With:
		b.WriteString("(with ")
		writeExpr(b, e.Scope)
		b.WriteString("; ")
		writeExpr(b, e.Body)
		b.WriteString(")")
	case Assert:
		b.WriteString("(assert ")
		writeExpr(b, e.Condition)
		b.WriteString("; ")
		writeExpr(b, e.Body)
		b.WriteString(")")
	case If:
		b.WriteString("(if ")
		writeExpr(b, e.Condition)
		b.WriteString(" then ")
		writeExpr(b, e.Then)
		b.WriteString(" else ")
		writeExpr(b, e.Else)
		b.WriteString(")")
	default:
		panic(fmt.Sprintf("unknown expression %T", e))
	}
}

// ToNix renders the AST as Nix source.
func ToNix(e Expr) string {
	var b strings.Builder
	b.Grow(512)
	writeExpr(&b, e)
	return b.String()
}
